sdk/program/src/vote/state/mod.rs

```diff
@@ -927,9 +928,11 @@ impl VoteState {
     // allows validators to switch forks once their votes for another fork have
     // expired. This also allows validators continue voting on recent blocks in
     // the same fork without increasing lockouts.
-    pub fn pop_expired_votes(&mut self, next_vote_slot: Slot) {
+    pub fn pop_expired_votes(&mut self, next_vote_slot: Slot, parent_slot: u64) {
         while let Some(vote) = self.last_lockout() {
-            if !vote.is_locked_out_at_slot(next_vote_slot) {
+            if !vote.is_locked_out_at_slot(next_vote_slot)
+                && (parent_slot == 0 || vote.slot != parent_slot)
```

Comment:
This violates the consensus protocol, see https://docs.solanalabs.com/implemented-proposals/tower-bft#vote-tower for an overview of the specification. Also requires a feature flag.

If you wish to suggest a change to the consensus protocol please follow the instructions to submit a SIMD https://github.com/solana-foundation/solana-improvement-documents/blob/main/proposals/0001-simd-process.md.

Reply:
Can you point out which part of the specification is violated?  The specification is definitely out-of-date as the VoteStateUpdate implementation does not require the expirations that this specification describes.  If the violation you are referencing is that part of the spec, then the standard code in Anza's/Solana Labs' implementation is also violating the spec.

Reply:
This requires a feature flag since it would change behavior/credits between a validator who implements the change and a validator that does not implement the change.    The concept of the change is good but there is just procedures that need to be followed if it has an effect on consensus.     From the firedancer perspective, this change feels easy and straightforward to mirror ...

Reply:
This part
![image](https://github.com/user-attachments/assets/02cb1f56-afc6-47b2-862f-690cb24d195b)

Agreed that our documentation is lacking, but this describes TowerBFT not the vote program.

Reply:
Regardless, such a big change to consensus should go through the SIMD process. Specifically we would like to see evidence that a change to lockouts are necessary and improve consensus performance without breaking liveness.

The change as written in this PR breaks liveness in situations where we are voting on a fork with little stake, so I believe more discussion is needed. Based on the title it seemed that this was a bug fix which it is not.

Reply:
Also from the description:
> resulting in missed credits even though they voted on valid slots. This results in 1-2% lower vote credits.

Is the main motivation behind this suggestion vote credits or issues with the consensus algorithm itself? If this is motivated by vote credits I would suggest exploring changes to the credit algorithm first as it is easier to modify and reason about.

Reply:
I don't understand how this would break liveness due to lack of deep understanding of the network wide consensus and edge cases. It's important to acknowledge that now the information is public, an increasing number of validator operators will begin deploying similar changes to increase their credits and APY. If there is significant risk to the cluster, it would be valuable to communicate exactly how it would create a liveness issue to disincentivize anyone from running a similar change.

Regarding a feature flag, I think that's fine. I'm still not certain whether this implementation, @bji's, or another is the best for the network.

> Is the main motivation behind this suggestion vote credits or issues with the consensus algorithm itself?
Vote credits. As mentioned in discord, a validator operator began advertising their services to other operators with the selling point that they _know the secret_ of how to get increased vote credits. This prompted @bji to go public as one of the few with knowledge of how to increase vote credits.